File: src/main/java/cc/sukazyo/cono/morny/bot/query/MornyQueries.java
package cc.sukazyo.cono.morny.bot.query;

import cc.sukazyo.cono.morny.bot.api.InlineQueryUnit;
import com.pengrad.telegrambot.model.Update;

import javax.annotation.Nonnull;
import java.util.ArrayList;
import java.util.List;

public class MornyQueries {
	
	private final List<ITelegramQuery> queryInstances = new ArrayList<>();
	
	public MornyQueries () {
		queryInstances.add(new RawText());
		queryInstances.add(new MyInformation());
		queryInstances.add(new ShareToolTwitter());
	}
	
	@Nonnull
	public List<InlineQueryUnit<?>> query (@Nonnull Update event) {
		final List<InlineQueryUnit<?>> results = new ArrayList<>();
		for (ITelegramQuery instance : queryInstances) {
			final List<InlineQueryUnit<?>> r = instance.query(event);
			if (r!=null) results.addAll(r);
		}
		return results;
	}
	
}

File: src/main/scala/cc/sukazyo/cono/morny/bot/query/MornyQueries.scala
package cc.sukazyo.cono.morny.bot.query

import cc.sukazyo.cono.morny.bot.query
import cc.sukazyo.cono.morny.MornyCoeur
import com.pengrad.telegrambot.model.Update

import scala.collection.mutable.ListBuffer

class MornyQueries (using MornyCoeur) {
	
	private val queryInstances = Set[ITelegramQuery](
		RawText(),
		MyInformation(),
		ShareToolTwitter(),
		ShareToolBilibili(),
		ShareToolXhs(),
		ShareToolSocialContent()
	)
	
	type QueryListenerExceptionListener = (ITelegramQuery, Throwable)=>Any
	
	def query (onQueryListenerExceptions: QueryListenerExceptionListener)(event: Update): List[InlineQueryUnit[_]] = {
		val results = ListBuffer[InlineQueryUnit[_]]()
		for (instance <- queryInstances) {
			try {
				val r = instance query event
				if (r != null) results ++= r
			} catch case e => onQueryListenerExceptions(instance, e)
		}
		results.result()
	}
	
}
